### ISSUE TITLE
Text scaling strangely when DPI-scaling is changing

### ISSUE THREAD

**Issue by Naouak** [mintty/mintty#492]
Hi,

On my Surface Pro 4, the application is scaling well on the SP4 screen(200% dpi setting). If I switch the application to a second screen with 100% dpi setting, the status bar is not reduced and appear awfully big and the text is very small (unreadable).

If I deactivate DPI scaling for the application, the status bar is still problematic but the text is fine.

Thanks,

**Comment by mintty** [comment]
Which versions (mintty, OS) are you running?
Is there a difference between mintty 2.1.5 (or older) and 2.2.1 (or newer)?
If you hold Shift while dragging the window to the other screen, what's the difference?
Can you provide screenshots?

**Comment by Naouak** [comment]
I'm using 2.2.2 on Win 10 (with TH2 update) (version grabbed from Cygwin X64 this morning).
![mintty](https://cloud.githubusercontent.com/assets/150214/11147557/fff69d3e-8a16-11e5-9c75-b6d14e56a54b.png)
This is one screenshot. It's not easy to see on  On the top is my low DPI screen, on the bottom the high DPI (please refer to start menu for proportions).

Holding Shift while dragging doesn't seem to make any difference.

**Comment by mintty** [comment]
So you drag an instance of the lower window to the other screen where it looks like the upper window?
The upper one is apparently in full-screen mode (telling by its title bar symbols), can you confirm?
Symbols are cut off for the lower one, is it full-screen too?
What does `stty size` report in both?
Issue is similar to #487, maybe a duplicate.

**Comment by Naouak** [comment]
both are maximized on different screens. None of the Windows UI is scaled (look at the scroll bar and status bar, context menu is also not scaled)

Here is `stty size` values:
TOP : 124 376
Bottom : 70 192

**Comment by mintty** [comment]
So you drag a maximised window and it stays maximised? (This is different in Windows 7 so I cannot reproduce the exact issue right now.)
What happens if you use `Win`+`Shift`+cursor-left/right instead of dragging?

**Comment by Naouak** [comment]
When dragging without maximizing, the font is still reducing but when using window snapping shortcuts it keeps the font size. If I drag the window from bottom screen to top screen, when DPI-scaling kicks in, the font is reduced. 

It seems there is a multiplication of the font size that is inconsistent when switching screens. If I switch windows several times, the fonts keep getting smaller.
If I use snapping to get to the top screen, the screen keep the same pixel size and if I drag it down it seems to grows.

Maybe the issue is with the way the sacling factor is calculated and when it is applied ?

**Comment by Naouak** [comment]
The UI keeping the same size is another issue I guess since it's always consistant pixel wise and does not scale according to scaling parameters.

**Comment by mintty** [comment]
Try to set option ZoomFontWithWindow=false s’il vous plaît.

**Comment by mintty** [comment]
The title bar is handled by Windows, not mintty, by the way. This isn't likely to change as a self-made title bar would be much effort. I'd like to close that aspect of the issue and remove it from the issue title.

**Comment by Naouak** [comment]
Strangely enough for the Title Bar, every other application is scaling well. Maybe there is a flag on the application that tells windows to not scale the UI ? cmd.exe for example is scaling correctly the title bar and scroll bar.

Problem is still there with this .minttyrc:

    BoldAsFont=no
    Transparency=low
    ZoomFontWithWindow=false

**Comment by espresso3389** [comment]
Naouak,

> When dragging without maximizing, the font is still reducing but when using window snapping shortcuts it keeps the font size. If I drag the window from bottom screen to top screen, when DPI-scaling kicks in, the font is reduced.

I guess your display layout is like the following one. Is it correct?

![layout](https://cloud.githubusercontent.com/assets/1311400/11152446/e1b87482-8a75-11e5-9cf5-89931fa615e0.png)

I have no idea why such issue occurs. But anyway I will try to reproduce the issue on my environment.

**Comment by Naouak** [comment]
Exactly.

**Comment by espresso3389** [comment]
When we handle WM_DPICHANGED, we assume that WM_DPICHANGED's lParam is a RECT, which contains the window location on the new monitor calculated from the original window location on the previous monitor. It is, if the window was 96x96 (96dpi) on the previous monitor, we expected that lParam contains 144x144 on 144dpi monitor.
But, in reality, if the Window is moved using ways like Aero-snap, lParam contains the size of window size snapped to the monitor and it is not 144x144 in such case.

And, we can detect whether the window is aero-snapped by the following code (https://github.com/espresso3389/mintty/blob/master/src/winmain.c) :
```cpp
static bool
is_resize_by_aerosnap(HWND wnd)
{
  // The function works because of the fact that:
  // When the window is resized by Aero-snap, GetWindowPlacement returns
  // SW_SHOWNORMAL but wp.rcNormalPosition is not identical to the size
  // returned by GetWindowRect.
  WINDOWPLACEMENT wp;
  wp.length = sizeof(WINDOWPLACEMENT);
  GetWindowPlacement(wnd, &wp);
  if (wp.showCmd == SW_SHOWMAXIMIZED)
    return true;
  if (wp.showCmd != SW_SHOWNORMAL)
    return false;
  RECT rc;
  GetWindowRect(wnd, &rc);
  return memcmp(&rc, &wp.rcNormalPosition, sizeof(RECT)) != 0;
}
```

But even with the code, we still could not properly determine all the cases...

**Comment by mintty** [comment]
Is this affected by the following setting?
Control Panel – Display – Set custom text size – Use Windows XP style DPI scaling

**Comment by espresso3389** [comment]
No, not at all. They are completely different things.

**Comment by AlexSamad** [comment]
Hi

I have a Dell XPS 13
laptop res = 3200x1800
t displays attached = 2560 x 1440 

my default screen is my laptop 
opening on the laptop screen seems to be okay - border and text size.

I have uploaded a screenshot of 2 windows that have been dragged from the laptop screen to the external screens.  the tiny font one is dragged with shift held down. and the larger text is without shift being held.

$ mintty --version
mintty 2.3.7 (x86_64-pc-cygwin)
(C) 2013/2016 Andy Koppe / Thomas Wolff
License GPLv3+: GNU GPL version 3 or later
There is no warranty, to the extent permitted by law.


![image](https://cloud.githubusercontent.com/assets/5509219/15593124/c2c8fe50-23ec-11e6-9092-0043282f2bd5.png)

**Comment by jingsam** [comment]
Windows 10 mintty 2.3.7, same problem when I drag mintty to another screen.

**Comment by mintty** [comment]
I have the impression that different scenarios are being reported here. Please describe precisely what the issue is.

In @AlexSamad's upper screenshot, what's the problem.
In the lower screenshot, what does `stty size` report before and after moving? Does option `AnywayZoomFontWithWindow=false` prevent the font from becoming tiny? (This is a minor issue, no need to use Shift when moving the window.)

**Comment by AlexSamad** [comment]
Hi

The top windows is massive both of those are 9point.


![image](https://cloud.githubusercontent.com/assets/5509219/15729427/73ac4894-28a7-11e6-8f32-db282760929a.png)


This might be better.
Back window - normal windows started on my laptop (where it sort of looks okay) dragged to external window
front left is normal windows dragged with shift to external window
front right it git-bash window - its the only one that is actually working well.

they are all 9point lucinda console

**Comment by mintty** [comment]
To get an impression of the visual mismatch that you seem to see in the "back window", can you take a picture (not screenshot) of both monitors running the same window configuration on each?
This is probably something that can be tweaked by detecting the actual DPI value of each monitor.
Also, please state the resolution and the DPI values / scaling factors you configured for both monitors.

**Comment by AlexSamad** [comment]
left and right monitors are the same 
DELL U2715H = 2560 x 1440
laptop = 3200 x 1800

left windows is git-bash, started on laptop and moved
right is windows that started on the laptop and was full sized to the window on the right
window on laptop started there and got full sized

all lucinda 9pt 

![img_20160603_093315](https://cloud.githubusercontent.com/assets/5509219/15764553/fc4ccbe8-296e-11e6-8053-849a033fc49d.jpg)

**Comment by mintty** [comment]
What are the zoom factors of all monitors?
Right-click desktop, look at slider (values 100%, 125%, ...), click on each monitor image, what does the slider value say?

**Comment by mintty** [comment]
Actually, I can make the whole DPI-dependent adaptation handling optional.
This should affect both title bar and font size; however, it's questionable whether that's the proper thing to do.

Let's consider your monitor configuration. Unfortunately, you didn't tell me the configured DPI values, so let's assume they are equal. Then your systems sees two large monitors (2560 x 1440) and one even larger monitor (3200 x 1800). When a window is moved from the largest (the laptop monitor, as the system cannot know that it's in fact much smaller) to a smaller one, the font should be scaled up to compensate for that and achieve visually similar size after moving. This behaviour is actually correct.

**Comment by AlexSamad** [comment]
Hi

Not infront of my pc at the moment.

Making it optional sounds good.

But ... git-bash handles it okay ...  The font size is good the title bar
is good.

But I do miss the rxvt-native - it worked well as well

A

**Comment by mintty** [comment]
Git-bash uses an outdated version of mintty that did not yet have the (proper!) DPI handling.

**Comment by mintty** [comment]
Option to ignore DPI changes released in 2.4.0.

**Comment by bmoyet** [comment]
Hi, 

I tried installing the new version, and setting the new parameter to false with no success. I edited the file .minttyrc in my home directory adding the following line : 

HandleDPI=false

It didn't change the bad behavior when i move cygwin on my other screen. The window title is still twice as big as it should be, and the font size is huge.

Just to be sure mintty is using the config file, I tried adding the parameter ZoomMouse=no in the config file, and it stopped the mouse zoom from working

**Comment by mintty** [comment]
If you start mintty from a desktop shortcut (if you don't, please try), in the shortcut properties, "Compatibility" tab, there is an option "Disable display scaling on high DPI settings". Any change?

**Comment by bmoyet** [comment]
It doesn't have any effect. The window title is still buggy

**Comment by mintty** [comment]
OK, same questions to you as to other reporters to this and similar issues:
Which Windows version?
What monitors do you have (pixel dimensions, physical sizes, individual DPI settings/zoom factors)?

**Comment by bmoyet** [comment]
My information:

Windows 10 Home
One laptop screen 4K Ultra HD (15,6") 3840 x 2160 zoom 250%
One old hdmi attached screen (22") 1920 x 1080 zoom 100%

**Comment by mintty** [comment]
4K @ 15" – that's crazy, and asking for trouble.
Anyway, you have the zoom factor which kind of compensates for the DPI gap, and slightly more, but not enough to bridge the physical size difference, too. Thus I would expect if you move mintty from the laptop screen to the external screen, its size should increase by some 20-40% (would have to calculate more thoroughly for a prediction).

**Comment by bmoyet** [comment]
If it can help, windows 10  is supposed to resize the window once the window is fully the external screen. At least it's what is happening with all programs I'm using. While moving the window and while it's on both screen the window appears big on my external screen. But ass soon as it's completely on the external screen, the window is scaled down at the right size. 

Except with mintty which stays the same. 

Two screenshot to show you the expected behavior.

![dual_before](https://cloud.githubusercontent.com/assets/7815879/16802057/c3d4de72-48fe-11e6-8317-641538fa6415.png)

![dual_after](https://cloud.githubusercontent.com/assets/7815879/16802063/c87dc894-48fe-11e6-95b7-bb4360da2a22.png)

**Comment by mintty** [comment]
DPI change handling tweaked in 2.4.1. Please recheck.

**Comment by bmoyet** [comment]
I tried this new version and see no changes. With or without the HandleDPI option.

I tried to debug dpi if it can helps. Here is the debug output when i open and move the window on my laptop screen (zoom 250%) :

> load_shcore_funcs shc 1
> SetProcessDpiAwareness 1 GetProcessDpiAwareness 1 GetDpiForMonitor 1
> per_monitor_dpi_aware 1
> monitor size 346mm*194mm res 3840*2160 dpi 240 eff 240

And here is the output and I drag the window onto my other screen and resize it twice (zoom 100%) :

> load_shcore_funcs shc 1
> SetProcessDpiAwareness 1 GetProcessDpiAwareness 1 GetDpiForMonitor 1
> per_monitor_dpi_aware 1
> monitor size 346mm*194mm res 3840*2160 dpi 240 eff 240
> monitor size 346mm*194mm res 3840*2160 dpi 240 eff 96
> monitor size 346mm*194mm res 3840*2160 dpi 240 eff 96
> monitor size 346mm*194mm res 3840*2160 dpi 240 eff 96

**Comment by mintty** [comment]
If you enable
```
#define debug_dpi
```
in line 128 of `winmain.c`, there should be more debug output than the one quoted, especially directly from the handling of WM_DPICHANGED. Your output looks as if it would scale 3 times, is this for multiple mouse actions, or for one move to another monitor only? I have not seen such behaviour in my testing. Maybe you could also enable 
```
#define debug_messages
```
in line 1104.

**Comment by mintty** [comment]
I seem to have misread your comment. So you explicitly resized twice in addition? Why? What's the visual effect before this?
For a more precise picture, can you please do this:
Enable the two debugs above.
Open two windows on the high-dpi monitor.
Move one of them over to the low-dpi monitor and do nothing else.
Capture the debug output of only the movement of only one window.
Take a picture (not screenshot) of both windows on both monitors, please.

**Comment by bmoyet** [comment]
Indeed I resized the window both before closing it in the second case. I can't enable **debug_messages** because I need _wm.t which I don't have. Maybe you can help me on this one. It may explains why I don't see any debug output. I have to launch mintty from cygwin with the following command, and then I have access to the outpu in the log file

> mintty > debug.log

The picture of my high dpi screen
![img_20160725_154429](https://cloud.githubusercontent.com/assets/7815879/17103550/1a056440-527f-11e6-97ab-c510c1687846.jpg)

The picture of my low dpi screen
![img_20160725_154150](https://cloud.githubusercontent.com/assets/7815879/17103565/2c473692-527f-11e6-9911-1f500a7aee40.jpg)

**Comment by mintty** [comment]
```
make _wm.t
```
OK, I had one picture of both screens in mind, for better visual size comparison, but calculating the physical and virtual dpi values, the window should be roughly 13% larger on the low-dpi screen. It looks about twice as large instead. This is certainly not expected.

**Comment by bmoyet** [comment]
Still impossible to make

>$ make _wm.t
>make: *** No rule to make target '_wm.t'.  Stop.

It's almost 250% larger, as if no transformation is applied to the cygwin window

![image](https://cloud.githubusercontent.com/assets/7815879/17106720/bda78c06-528c-11e6-9985-53d0addf6b86.png)

**Comment by mintty** [comment]
```
cd src; make _wm.t
```
I'm not sure if I can debug this properly as long as I don't have a similar setup myself.
In my tests with two monitors of different virtual dpi ("zoom"), but same physical size and dpi, it behaves well.
If anyone has a clue, your insights are appreciated.

**Comment by bmoyet** [comment]
Here are the complet output for the two case.

Just open mintty and close it :

[highdpi.txt](https://github.com/mintty/mintty/files/382003/highdpi.txt)

Open mintty, drag it to the lowdpi screen, and close it :

[lowdpi.txt](https://github.com/mintty/mintty/files/382004/lowdpi.txt)

**Comment by mintty** [comment]
Thanks a lot. It's quite weird that there is no WM_DPICHANGED message in the logs.
In a Windows shortcut, it is possible to deactive DPI scaling. Do you happen to have that activated?
Or do you have any mintty.exe.manifest file from previous testing?

**Comment by c-sonntag** [comment]
Try in the _.minttyrc_ this option  : 
`HandleDPI=false`

**Comment by mintty** [comment]
In the logs, DPI handling is apparently not invoked. In that case, `HandleDPI` makes no difference.

**Comment by bmoyet** [comment]
Indeed the mintty.exe had the option checked to disable dpi scaling... Now the window and the font size is somewhat adapted on my second screen. The only remaining problem is the window title bar which is still huge.

Here are the debug logs with the setting deactivated :

[lowdpi.txt](https://github.com/mintty/mintty/files/383230/lowdpi.txt)

**Comment by mintty** [comment]
I've tentatively revised DPI change handling to hook into WM_WINDOWPOSCHANGED rather than WM_DPICHANGED.
Please test the repository version (not making a release now before 3-week travelling).

**Comment by bmoyet** [comment]
Thanks,

I'll try tomorrow morning

**Comment by bmoyet** [comment]
It's actually worse than before. The window opens with huge character on my high dpi screen, and don't get resize at all when moving to the other screen.

Thanks for the try and have a nice vacation

**Comment by mintty** [comment]
Can you take another picture please? One picture with both screens, each with mintty as started or moved. What is your configured FontHeight?

**Comment by bmoyet** [comment]
After a few tries it seems the window gets resized after all, sorry about that. It seems it depend on my mouse position when a drag the window. I had no luck this morning.

But as you can see the font is big by default and the window title bar isn't resized at all. My FontHeight is 9pt I believe. If I set it to 4pt it's good on both screen.

![image](https://cloud.githubusercontent.com/assets/7815879/17364633/8fc7fd7e-5981-11e6-8617-8d74ae890603.png)

The only bug is still the title bar which don't get resized.

**Comment by mintty** [comment]
This picture confirms my estimation above that font display on the low-dpi screen should be ~13% larger. Also I guess the size roughly corresponds to 12pt but I'll check. We could discuss a constant scaling to approximate actual "pt" measurement. Another option you may try is to use a negative font size, which addresses some alternative coordinate system used in the Windows API.
Note that the title bar is a different issue.

**Comment by mintty** [comment]
Considering the following patch, wintext.c line 407:
```
    size > 0 ? -MulDiv(size, GetDeviceCaps(dc, LOGPIXELSY) * dpi, 72 * 96) : -size;
    size > 0 ? -MulDiv(size, GetDeviceCaps(dc, LOGPIXELSY) * dpi, 90 * 96) : -size;
```
in order to adjust the font size measure applied in mintty to that conceived by notepad.

**Comment by SCWells72** [comment]
I'm curious to see where this goes as well.  I also have a Surface Pro 4 that I dock to a 27" QHD screen at 100% scaling when I'm at home.  If I log into Windows on my home config then undock the SP4 (200% scaling undocked), most windows scale fine, but mintty text is so tiny as to be unreadable.  That's the case even if I close and reopen the window, so it's not just related to event handling.  I've tried both values for HandleDPI in .minttyrc, but that doesn't seem to make any difference.

I'm happy to provide any diagnostic info that might help here!

**Comment by mintty** [comment]
Checking different systems (still to be continued), the patch should rather look like:
```
-    size > 0 ? -MulDiv(size, GetDeviceCaps(dc, LOGPIXELSY) * dpi, 72 * 96) : -size;
+    size > 0 ? -MulDiv(size, dpi, 72) : -size;
```
You are welcome to try this yourself, or wait ~10 days for an upload.

**Comment by bmoyet** [comment]
Using this patch, when I open a new cygwin the font is too small.

Should I open another issue for the title bar ?

**Comment by mintty** [comment]
What do you mean by "too small"? Like a little bit smaller?
Depending on the screen dimensions, the font may be smaller than it used to be but that is actually part of the solution. This is the price for finally fixing this issue. Please compare the "new" font size with the respective font size of other applications, e.g. notepad; it should align with that. I will guard the changed behaviour with the existing option `HandleDPI=false`.

For the title bar, there is an issue already. I think I closed it though, because it's actually a Windows design bug.

**Comment by mintty** [comment]
[back in 10 days]

**Comment by bmoyet** [comment]
My bad, I didn't removed the font size I set the last time.

It seems to be very close now... A little bit smaller maybe, like 2 pixels...

![image](https://cloud.githubusercontent.com/assets/7815879/17615921/ae711930-6071-11e6-83f2-6927e3d02736.png)

**Comment by mintty** [comment]
Revised DPI handling as described in https://github.com/mintty/mintty/pull/470#issuecomment-241581145; release in 2.5.0 (cygwin test version).

**Comment by SCWells72** [comment]
I can confirm that the main window in 2.5.0 is a major improvement on my Surface Pro 4.  Thanks!  The window frame and options dialog are still tiny.  Any change of having those react to the same scaling information?  If so, I think that will take care of it.  Either way, this is shaping up very nicely!

**Comment by mintty** [comment]
Unfortunately, the options dialog font size is configured statically, in `res.rc`. If you find out how to adjust that dynamically, that'll be great. Interference of font size and dialog items geometry is to be expected, unless the whole thing can be handled in some kind of logical coordinates and scaled. See `windialog.c`.

**Comment by SCWells72** [comment]
It's been a LONG time since I've done any Windows programming, so take this with a grain of salt, but perhaps this might help:

http://www.codeproject.com/Articles/5451/Fight-the-dialog-units-DPI-and-Large-Fonts

I don't know if you could use that when fielding your WM_INITDIALOG message to set the DPI according to the scaling factor of the display.

**Comment by mintty** [comment]
Thanks for the interesting reference. I have no idea, though, how to apply that to the mintty code.

It turns out on the other hand, that at least the pixel geometry is automatically scaled with the font size, i.e. if a different font size is specified in res.rc, the dialog window adjusts properly. This seems to be achieved with `MapDialogRect(wnd, &r);` after `WM_INITDIALOG`.

Now we should find out how to scale the dialog font size _dynamically_. Maybe by fiddling some magic between messages `WM_GETFONT` and `WM_SETFONT`? But then the order of this and `MapDialogRect` is rather obscure, which should likely depend on the scaling factor. What does this crazy `MapDialogRect` consider anyway for scaling?

**Comment by SCWells72** [comment]
Okay, I believe I've figured out how to dynamically change the font for the dialog controls.  Add the following to `windialog.c`:

```
static bool CALLBACK set_dialog_font(HWND child, LPARAM font) {
  SendMessage(child, WM_SETFONT, font, true);
  return true;
}
```
and the following to the beginning of your `WM_INITDIALOG` handler:

```
      // TODO: Get the font from the resource and adjust the point size (24 here)
      // TODO: according to the scaling factor.
      HFONT dialogFont = CreateFont(24, 0, 0, 0, FW_BOLD, FALSE,
        FALSE, FALSE, ANSI_CHARSET, OUT_DEFAULT_PRECIS,
        CLIP_DEFAULT_PRECIS, DEFAULT_QUALITY, DEFAULT_PITCH,
        "MS Shell Dlg");
      EnumChildWindows(wnd, (WNDENUMPROC) set_dialog_font, (LPARAM) dialogFont);
```

I've hard-coded 24 there and the dialog is completely based on MS Shell Dlg 24 point.  As noted in the comment, you'd want to get the window's current font and scale the point size according to the scaling factor.

**Comment by mintty** [comment]
Thanks; I've added this to the beginning of `config_dialog_proc`:
```
  bool CALLBACK set_dialog_font(HWND child, LPARAM font) {
    SendMessage(child, WM_SETFONT, font, true);
printf("setting dialog font\n");
    return true;
  }

  void adustdialogfontsize()
  {
    int diafontsize = MulDiv(cfg.font.size, dpi, 96);

    HFONT dialogfont = CreateFont(diafontsize, 0, 0, 0, FW_DONTCARE, 
      false, false, false, 
      DEFAULT_CHARSET, OUT_DEFAULT_PRECIS,CLIP_DEFAULT_PRECIS, 
      DEFAULT_QUALITY, DEFAULT_PITCH,
      "MS Shell Dlg");
printf("diafontsize %d diafont %p\n", diafontsize, dialogfont);
    EnumChildWindows(wnd, (WNDENUMPROC) set_dialog_font, (LPARAM) dialogfont);
  }

```
and inserted `adustdialogfontsize();` at various positions; it needs to be call towards the end of handling `WM_INITDIALOG`, but still facing problems:
* It affects only the first Options view; as soon as you change categories, the old font size will reappear (probably easy to fix)
* It does not reuse the res.rc font configuration `FONT 8, "MS Shell Dlg"`; easy to fix, this could be a macro like DIALOG_HEIGHT
* It does not scale the dialog dimensions, as would be configured statically in res.rc `FONT 8`...

So there is still some investigation to be done...

**Comment by mintty** [comment]
It might be useful to check `RegisterClass` in `win_open_config`, and change the static value in `.lpszClassName = "ConfigBox"`. But I have no idea how a class and new class name could be defined with a dynamic geometry and font size. (As a workaround, a range of predefined classes for various typical scaling factors could be used...)

**Comment by SCWells72** [comment]
Bummer that the simple version didn't work, but not surprising I guess.

Yeah, I was thinking that a canned set of predefined classes would do it.  If you check MS's docs:

https://msdn.microsoft.com/en-us/library/windows/desktop/dn469266(v=vs.85).aspx

and look at "DPI and the Desktop Scaling Factor", there are only really a few scaling factors and well-defined associated DPIs:

96 DPI = 100% scaling
120 DPI = 125% scaling
144 DPI = 150% scaling
192 DPI = 200% scaling

My guess is that if you provided canned versions of these, you would pretty much address it.

**Comment by SCWells72** [comment]
I don't know if there's anything useful here, but this seems to be MS's latest-and-greatest article on display scaling as of Windows 10 Anniversary Update:

https://blogs.technet.microsoft.com/askcore/2016/08/16/display-scaling-changes-for-the-windows-10-anniversary-update/

**Comment by mintty** [comment]
That link is very interesting (thanks), and confirms my previous comments about the MS crap...
Maybe it's possible to switch the DPI handling mode dynamically before opening an Options dialog;
I'll try that first before crafting that workaround.

**Comment by SCWells72** [comment]
Okay...and seriously don't spend too much time on this.  The big thing was getting the terminal window itself rendering properly.  My guess is that people don't spend much time configuring options once it's set up the way you want, and this is really only an issue when the DPI has changed since logging in.  Thanks so much for the support!

**Comment by mintty** [comment]
Thanks for caring but isn't the Options menu on a high-DPI monitor simply too tiny, without having changed the DPI?

**Comment by mintty** [comment]
Another link I found describes the issue well, including the interference of blurriness and title bar scaling:
http://stackoverflow.com/questions/36864894/scaling-the-non-client-area-title-bar-menu-bar-for-per-monitor-high-dpi-suppo/37624363

**Comment by SCWells72** [comment]
Yes, the context menu, Options dialog, and title bar are all tiny on a high DPI display when the scaling factor is mismatched.  To be specific, when I'm docked at home, I have a 19" 1280x1024 monitor, a 24" 1920x1080 monitor, a 27" 2560x1440 monitor, and the 12" 2736x1824 screen of the Surface Pro 4 all active (yes, I like having plenty of screen real estate!).  All of the external monitors use 100% scaling and the SP4 screen uses 200% scaling.  These are all Windows' recommended scaling factors of these screens, so I haven't done anything abnormal here.

The 27" WQHD screen is set as primary, so the scaling factor for apps at creation time is 100%.  When the mintty window is on any of the external monitors, everything looks fine.  When I move it to the SP4 display, the terminal text is sized properly now but the context menu, Options dialog, and title bar are all too small to be usable.  This is without ever docking/undocking.

Let me know if there's anything else I can do to help.  Again, I definitely appreciate the support!

**Comment by mintty** [comment]
With the new API in the Windows 10 Anniversary Update it is possible to have the Options menu auto-scale while scaling of the terminal window remains mintty-driven. This brings back the font blurring issue to us but only for the Options menu where it may be acceptable.
Please check the current repository version (debug output still enabled).
I have also implemented enabling of auto-scaling of the title bar as documented but it does not work...

**Comment by mintty** [comment]
Released tweak of previous fix for Windows 7/XP in 2.5.1.
Also released DPI scaling of Options menu.
Intending to close this issue.

**Comment by SCWells72** [comment]
I can confirm that the options dialog and terminal text are both readable now where they weren't before.  Thanks so much for the work!

**Comment by mintty** [comment]
Released revised DPI handling in 2.5.1.

**Comment by mintty** [comment]
The current set of tweaks has the unpleasant side effect that (in Windows 10 with monitor zooming factor > 100%) on font selection (from the Options menu) the terminal gets resized by the zooming factor.
Therefore I need to revert part of the tweaks.

On the other hand, it turns out that with the recent Windows 10 update, most of them are not necessary.
Please re-check the current version.

**Comment by SCWells72** [comment]
Okay.  It'll likely be Monday before I get a chance to take a look.  I'll let you know!

**Comment by mintty** [comment]
After one more tweak (adjusting window size) that should not affect font scaling, I hope this issue is now resolved in a stable way.

**Comment by mintty** [comment]
Released 2.6.1.

**Comment by SCWells72** [comment]
Just tried out 2.6.1 on my Surface Pro 4 that's undocked after having scaling factor set while docked.  The terminal text and menu bar both look perfect.  The context menu and options dialog are both tiny again.  NOT at problem at all, though, as it's VERY rare that I would need to go into those.  Those are almost certainly fine if I log out and back in even on Hi-DPI displays, though.  Overall I'm very happy with the current state.  Thanks!

**Comment by thelema** [comment]
I just updated to mintty 2.6.1, and am very happy to no longer need to adjust font sizes whenever I move mintty to a lower-DPI display. Thank you very much!

**Comment by mintty** [comment]
Note that failure of the context menu to scale properly is also mentioned elsewhere (http://stackoverflow.com/questions/36864894/scaling-the-non-client-area-title-bar-menu-bar-for-per-monitor-high-dpi-suppo, look for `CreatePopupMenu`). Hopefully MS will fix this.

**Comment by mintty** [comment]
With Windows DPI handling V2 (#774), now both Options and context menu are scaled properly.

**Comment by mintty** [comment]
Released 2.9.0.

**Comment by mintty** [comment]
Released 2.9.5 with default HandleDPI=2.
